nominal/ts.py

```diff
@@ -0,0 +1,238 @@
+EPOCH_MICROSECONDS = Epoch("microseconds")
+EPOCH_MILLISECONDS = Epoch("milliseconds")
+EPOCH_SECONDS = Epoch("seconds")
+EPOCH_MINUTES = Epoch("minutes")
+EPOCH_HOURS = Epoch("hours")
+
+_LiteralTimeUnit: TypeAlias = Literal[
+    "nanoseconds",
+    "microseconds",
+    "milliseconds",
+    "seconds",
+    "minutes",
+    "hours",
+]
+
+_LiteralAbsolute: TypeAlias = Literal[
+    "iso_8601",
+    "epoch_nanoseconds",
+    "epoch_microseconds",
+    "epoch_milliseconds",
+    "epoch_seconds",
+    "epoch_minutes",
+    "epoch_hours",
+]
+
+_LiteralRelativeDeprecated: TypeAlias = Literal[
+    "relative_nanoseconds",
+    "relative_microseconds",
+    "relative_milliseconds",
+    "relative_seconds",
+    "relative_minutes",
+    "relative_hours",
+]
+
+TypedTimeDomain: TypeAlias = Union[Iso8601, Epoch, Relative, Custom]
+"""Strongly typed time domain types."""
+
+_AnyTimeDomain: TypeAlias = Union[TypedTimeDomain, _LiteralAbsolute, _LiteralRelativeDeprecated]
+"""All allowable time domain types, including string representations."""
+
+
+def _make_typed_time_domain(domain: _AnyTimeDomain) -> TypedTimeDomain:
+    if isinstance(domain, (Iso8601, Epoch, Relative, Custom)):
+        return domain
+    if not isinstance(domain, str):
+        raise TypeError(f"timestamp type {domain} must be a string or an instance of one of: {TypedTimeDomain}")
+    if domain.startswith("relative_"):
+        # until this is completely removed, we implicitly assume offset=1970-01-01 in the APIs
+        warnings.warn(
+            "specifying 'relative_{unit}' as a string is deprecated and will be removed in a future version: use `nm.timedomain.Relative` instead. "
+            "for example: instead of 'relative_seconds', use `nm.timedomain.Relative('seconds', start=datetime.now())`. ",
+            UserWarning,
+        )
+    if domain not in _str_to_type:
+        raise ValueError(f"string time domains must be one of: {_str_to_type.keys()}")
+    return _str_to_type[domain]
+
+
+def _time_unit_to_conjure(unit: _LiteralTimeUnit) -> ingest_api.TimeUnit:
+    return ingest_api.TimeUnit[unit.upper()]
+
+
+_str_to_type: Mapping[_LiteralAbsolute | _LiteralRelativeDeprecated, Iso8601 | Epoch | Relative] = MappingProxyType(
+    {
+        "iso_8601": ISO_8601,
+        "epoch_nanoseconds": EPOCH_NANOSECONDS,
+        "epoch_microseconds": EPOCH_MICROSECONDS,
+        "epoch_milliseconds": EPOCH_MILLISECONDS,
+        "epoch_seconds": EPOCH_SECONDS,
+        "epoch_minutes": EPOCH_MINUTES,
+        "epoch_hours": EPOCH_HOURS,
+        "relative_nanoseconds": Relative("nanoseconds", start=0),
+        "relative_microseconds": Relative("microseconds", start=0),
+        "relative_milliseconds": Relative("milliseconds", start=0),
+        "relative_seconds": Relative("seconds", start=0),
+        "relative_minutes": Relative("minutes", start=0),
+        "relative_hours": Relative("hours", start=0),
+    }
+)
+
+
+class _SecondsNanos(NamedTuple):
```

Comment:
naming nit: _SecondsNanos is a v confusing name... would `_TimestampNanos` be fine? kinda doesn't matter how it's stored right? alternatively, `_SecondsNanoOffset` would be clearer

Reply:
this is only intended for internal use, but it's a tuple holding (seconds, nanoseconds) - didn't feel that weird to me

Reply:
maybe im just too used to api layers where this is the norm lol

Reply:
yeah the equivalent type in our conjure is `Timestamp`, hence the confusion to me I suppose - doesn't matter too much for internal types, I just generally like names that represent what something "is" vs describing what the fields are. trust you if this is a common enough pattern to be clear to others

Reply:
I get what you're saying and don't necessarily disagree, but `SecondsNanoOffset` sounds even more confusing to me and idk how to disambiguate better when there's tons of other timestamp types. Maybe just `_InternalTimestamp` or something

Reply:
cool, happy to leave as is